src/functional/iterator/embedded.py

```diff
@@ -887,34 +966,56 @@ def get_ordered_indices(
                 res.append(elem[sparse_position_tracker[axis.value]])
                 sparse_position_tracker[axis.value] += 1
             else:
-                assert isinstance(elem, (int, np.integer, slice))
+                assert isinstance(elem, (int, np.integer, slice, range))
                 res.append(elem)
     return tuple(res)
 
 
-def _tupsum(a, b):
-    def combine_slice(s, t):
-        is_slice = False
-        if isinstance(s, slice):
-            is_slice = True
-            first = 0 if s.start is None else s.start
-            assert s.step is None
-            assert s.stop is None
-        else:
-            assert isinstance(s, numbers.Integral)
-            first = s
-        if isinstance(t, slice):
-            is_slice = True
-            second = 0 if t.start is None else t.start
-            assert t.step is None
-            assert t.stop is None
-        else:
-            assert isinstance(t, numbers.Integral)
-            second = t
-        start = first + second
-        return slice(start, None) if is_slice else start
+@overload
+def _shift_range(range_or_index: range, offset: int) -> slice:
+    ...
+
+
+@overload
+def _shift_range(range_or_index: IntIndex, offset: int) -> IntIndex:
+    ...
 
-    return tuple(combine_slice(*i) for i in zip(a, b))
+
+def _shift_range(range_or_index, offset):
+    if isinstance(range_or_index, range):
+        # range_or_index describes a range in the field
+        assert range_or_index.step == 1
+        return slice(range_or_index.start + offset, range_or_index.stop + offset)
+    else:
+        assert is_int_index(range_or_index)
+        return range_or_index + offset
+
+
+@overload
+def _range2slice(r: range) -> slice:
+    ...
+
+
+@overload
+def _range2slice(r: IntIndex) -> IntIndex:
+    ...
+
+
+def _range2slice(r):
```

Comment:
Same comment as above about adding type hints to the actual implementation of a function with overloads.